File: network/Matrix.py
class Matrix:
    def __init__(self, dim = [3, 3]):
        self.dim = dim
        self.data = [[0 for x in range(dim[0])] for y in range(dim[1])]

    def size(self):
        return self.dim

    def width(self):
        return self.dim[0]

    def height(self):
        return self.dim[1]

    @staticmethod
    def convert(jaggedArray):
        
        height = len(jaggedArray)
        width = len(jaggedArray[0])
        matrix = Matrix([width, height])

        for i in range(height):
            for j in range(width):
                matrix.data[i][j] = jaggedArray[i][j]

        return matrix

    def toJaggedArray(self):
        return self.data

    def __delitem__(self, key):
        del self.data[key]

    def __getitem__(self, key):
        return self.data[key]

    def __setitem__(self, key, value):
        self.data[key] = value

    def __iter__(self):
        return self.data.__iter__

    def __str__(self):

        size = str(self.dim[0]) + ", " + str(self.dim[1])
        output = "Matrix <" + size + "> [\n"
        for row in self.data:
            for val in row:
                output += str(val) + ", "

            output += "\n"

        output += "]"
        return output
    
    def __repr__(self):
        return self.__str__()

File: java/src/Matrix.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;

class Matrix<T> implements Iterable {

    private int width;
    private int height;
    private ArrayList<ArrayList<T>> data;

    public Matrix(int[] dim) {
        this(dim[0], dim[1]);
    }

    public Matrix(int width, int height) {
        this.width = width;
        this.height = height;
        this.data = new ArrayList();

        for(int i = 0; i < height; i++) {
            this.data.add(new ArrayList<T>());
        }
    }

    public int[] size() {
        return new int[]{width, height};
    }

    public int width() {
        return this.width;
    }

    public int height() {
        return height;
    }

    public void set(int i, int j, T data) {
        this.data.get(i).set(j, data);
    }

    public T get(int i, int j) {
        return this.data.get(i).get(j);
    }

    public static <T> Matrix convert(T[][] jaggedArray) {

        int height = jaggedArray.length;
        int width = jaggedArray[0].length;
        Matrix<T> matrix = new Matrix(width, height);

        for (int i = 0; i < height; i++) {
            for (int j = 0; j < width; j++) {
                matrix.set(i, j, jaggedArray[i][j]);
            }
        }

        return matrix;
    }

    public void flatten(T[] array) {

        ArrayList<T> vector = new ArrayList<T>();
        for (int y = 0; y < this.height(); y++) {
            for (int x = 0; x < this.width(); x++) {
                vector.add(this.get(y, x));
            }
        }

        vector.toArray(array);
    }

//    public Matrix multiply(double val) {
//
//        Matrix m = new Matrix(this.size());
//        for (int y = 0; y < this.height(); y++) {
//            for (int x = 0; x < this.width(); x++) {
//                m.set(y, x, (int) (this.get(y, x) * val));
//            }
//        }
//
//        return m;
//    }

    public void toJaggedArray(T[][] array) {
        for( int y =0; y < this.height; y++) {
            this.data.get(y).toArray(array[y]);
        }
    }

    @Override
    public Iterator iterator() {
        return this.data.iterator();
    }

    @Override
    public String toString() {

        String size = this.width + ", " + this.height;
        String output = "Matrix <" + size + "> [\n";
        for (ArrayList<T> row : this.data) {
            for (T val : row) {
                output += val + ", ";
            }

            output += "\n";
        }

        output += "]";
        return output;
    }
}